server/SchemaLoader.js

```diff
@@ -0,0 +1,171 @@
+        }.bind(this);
+        req.send();
+    }
+
+    storeSchemas(schemas) {
+        // Potentially it's only one schema, if so convert to array
+        ((Array.isArray(schemas) && schemas) || [schemas]).forEach((schema) => {
+            let content = JSON.stringify(schema);
+            let schemaPath = path.join(this.getSchemasPath(), this.getNameFromSchema(schema));
+
+            // Create folders recursively, if they don't already exist.
+            mkdirp(path.dirname(schemaPath), function(mkdirError) {
+                if (mkdirError) {
+                    console.log(mkdirError);
+                    return;
+                }
+
+                // Save schemas to disk. Always update, schemas might've been updated.
+                fs.writeFile(schemaPath, content, function(writeError) {
+                    if (writeError) {
+                        console.log(writeError);
+                        return;
+                    }
+                });
+            });
+        }, this);
+    }
+
+    getNameFromSchema(schema) {
+        // prettier-ignore
+        return [
+            schema.self.vendor,
+            schema.self.name,
+            schema.self.format,
+            schema.self.version
+        ].join("/");
+    }
+
+    getSchemasPath() {
+        let schemaDir = remote.getGlobal("options").schemaDir;
+        // in production, remove app.asar from the path
+        // cannot use process.resourcesPath in development,
+        // as that will point to electron/dist in node_modules
+        let resourcesPath = remote.app.getAppPath().replace("app.asar", "");
+        return !!schemaDir && path.join(resourcesPath, schemaDir);
+    }
+
+    loadLocalSchemas() {
+        let schemasPath = this.getSchemasPath();
+        let schemas = this.readSchema(schemasPath, {});
+        logger.info("Schemas loaded from disk");
+        this.emit("schemas-loaded", schemas);
+    }
+
+    readSchema(file, schemas) {
+        let stats;
+        try {
+            stats = fs.lstatSync(file);
+        } catch (err) {
+            // catch in case the file or directory could not be resolved,
+            // when e.g. somebody deleted the schemas folder
+            console.log(err);
+            return schemas;
+        }
+
+        if (stats.isFile() && (path.extname(file) === "" || path.extname(file) === ".json")) {
+            // only accept .json or no extension
+            try {
+                let schema = jsonfile.readFileSync(file);
+                let schemaName = this.getNameFromSchema(schema);
+                schemas[schemaName] = new ValidationSchema(schemaName, schema);
+            } catch (err) {
+                // catch non-valid JSON schemas
+                console.log(file + ": " + err);
+            }
+        } else if (stats.isDirectory(file)) {
+            let files = fs.readdirSync(file);
+            files.forEach(function(f) {
+                schemas = this.readSchema(path.join(file, f), schemas);
```

Comment:
`readSchema` doesn't "use" any previously collected schemas, so it seems possible to drop the second parameter; the code here would then update, rather than reassign, `schemas`. The result is a much more functional style of programming.

Reply:
You're completely right. Think I've fixed it now :)